File: examples-java/src/main/java/io/github/matthewjones372/proofload/examples/java/Checkout.java
package io.github.matthewjones372.proofload.examples.java;

import io.github.matthewjones372.proofload.RunResult;
import io.github.matthewjones372.proofload.Scenario;
import io.github.matthewjones372.proofload.SessionKey;
import io.github.matthewjones372.proofload.StepName;
import io.github.matthewjones372.proofload.Verdict;
import io.github.matthewjones372.proofload.http.Http;
import io.github.matthewjones372.proofload.java.Goals;
import io.github.matthewjones372.proofload.java.Https;
import io.github.matthewjones372.proofload.java.Proofload;
import io.github.matthewjones372.proofload.java.Rates;
import io.github.matthewjones372.proofload.java.Results;
import io.github.matthewjones372.proofload.java.Scenarios;
import io.github.matthewjones372.proofload.java.SessionKeys;
import io.github.matthewjones372.proofload.java.Simulations;
import io.github.matthewjones372.proofload.java.Steps;
import java.time.Duration;

/**
 * A load test written in Java, and the gate on {@code proofload-java} being
 * callable from it. Nothing here imports Kotlin and no name here carries a
 * value-class hash; a facade method that goes takes this source set with it.
 */
public final class Checkout {

    private static final SessionKey<String> ORDER_ID = SessionKeys.of(String.class, "orderId");

    private static final StepName BROWSE = Steps.named("browse");

    private static final StepName PLACE_ORDER = Steps.named("place order");

    private Checkout() {
    }

    public static void main(String[] args) {
        Http api = Https.baseUrl("https://orders.internal");

        Scenario checkout = Scenarios.named("checkout")
            .exec(BROWSE, api.get("/products"))
            .exec(PLACE_ORDER, Https.capturing(
                api.post("/orders").body("{\"cart\":\"1 anvil\"}").expecting(201),
                ORDER_ID,
                response -> response.header("location")))
            .pause(Duration.ofSeconds(1))
            .build();

        RunResult result = Proofload.create().run(
            Simulations.at(checkout, Rates.perSecond(50), Duration.ofMinutes(1),
                Goals.p99Under(PLACE_ORDER, Duration.ofMillis(200)),
                Goals.failureRateUnder(0.1)));

        for (Verdict verdict : Results.verdicts(result)) {
            System.out.println(verdict.getGoal().getDescribed() + (verdict.getMet() ? " met" : " missed"));
        }

        Duration tail = Results.p99(result, PLACE_ORDER);
        System.out.println(PLACE_ORDER.getName() + " p99 " + tail.toMillis() + "ms over "
            + Results.count(result, PLACE_ORDER) + " requests, " + Results.failed(result, PLACE_ORDER) + " failed");
    }
}

File: examples-scala/src/main/scala/io/github/matthewjones372/proofload/examples/scala/Checkout.scala
package io.github.matthewjones372.proofload.examples.scala

import io.github.matthewjones372.proofload.java.Https
import io.github.matthewjones372.proofload.java.Results
import io.github.matthewjones372.proofload.java.Simulations
import io.github.matthewjones372.proofload.scala.Proofload
import io.github.matthewjones372.proofload.scala.apply
import io.github.matthewjones372.proofload.scala.exec
import io.github.matthewjones372.proofload.scala.failureRate
import io.github.matthewjones372.proofload.scala.given
import io.github.matthewjones372.proofload.scala.http
import io.github.matthewjones372.proofload.scala.p99
import io.github.matthewjones372.proofload.scala.pause
import io.github.matthewjones372.proofload.scala.percent
import io.github.matthewjones372.proofload.scala.perSecond
import io.github.matthewjones372.proofload.scala.scenario
import io.github.matthewjones372.proofload.scala.sessionKey
import io.github.matthewjones372.proofload.scala.step
import _root_.scala.concurrent.duration.DurationInt
import _root_.scala.language.implicitConversions

/**
 * A load test written in Scala, and the gate on `proofload-scala` being callable
 * from it. No name here carries a value-class hash and no duration here is a
 * bare `Long`; an extension that goes takes this source set with it.
 */
object Checkout:

  private val orderId = sessionKey[String]("orderId")

  private val browse = step("browse")

  private val placeOrder = step("place order")

  def main(args: Array[String]): Unit =
    val api = http.baseUrl("https://orders.internal")

    val checkout = scenario("checkout")(
      exec(browse, api.get("/products").expecting(200)),
      exec(
        placeOrder,
        Https.capturing(
          api.post("/orders").body("""{"cart":"1 anvil"}""").expecting(201),
          orderId,
          response => response.header("location"),
        ),
      ),
      pause(1.second),
    )

    val result = Proofload().run(
      Simulations.at(
        checkout,
        50.perSecond,
        1.minute,
        p99(placeOrder) under 200.millis,
        failureRate under 0.1.percent,
      ),
    )

    Results.verdicts(result).forEach: verdict =>
      println(verdict.getGoal.getDescribed + (if verdict.getMet then " met" else " missed"))

    val tail = result(placeOrder).responseTime.p99
    println(s"${placeOrder.getName} p99 ${tail.toMillis}ms over ${result(placeOrder).count} requests, " +
      s"${result(placeOrder).failed} failed")
